In [ ]:
print("\n" + "=" * 80)
print("PHÁT HIỆN CỘNG ĐỒNG (COMMUNITY DETECTION)")
print("=" * 80)

# Sử dụng Louvain method (greedy modularity optimization)
print("\n[Louvain Method - Greedy Modularity Optimization]")
print("  → Phát hiện cộng đồng từ mạng...")

from networkx.algorithms import community

# Louvain (greedy modularity optimization)
communities_louvain = list(community.greedy_modularity_communities(G_undirected))
modularity_louvain = community.modularity(G_undirected, communities_louvain)

print(f"✓ Số cộng đồng phát hiện: {len(communities_louvain)}")
print(f"✓ Modularity Score: {modularity_louvain:.6f}")

# Phân tích cộng đồng
print("\n[Phân tích chi tiết cộng đồng]")
community_stats = []

for idx, comm in enumerate(communities_louvain):
    comm_nodes = list(comm)
    comm_size = len(comm)
    
    # Loại nodes trong cộng đồng
    node_types_in_comm = defaultdict(int)
    for node_id in comm_nodes:
        node_type = nodes_df[nodes_df['id'] == node_id]['type'].values[0] if node_id in nodes_df['id'].values else 'unknown'
        node_types_in_comm[node_type] += 1
    
    # Lấy top nodes theo degree
    subgraph = G_undirected.subgraph(comm_nodes)
    top_nodes_in_comm = sorted(subgraph.degree(), key=lambda x: x[1], reverse=True)[:3]
    
    community_stats.append({
        'community_id': idx,
        'size': comm_size,
        'percentage': (comm_size / G_undirected.number_of_nodes() * 100),
        'node_types': dict(node_types_in_comm),
        'top_nodes': top_nodes_in_comm
    })
    
    print(f"\nCộng đồng {idx + 1}:")
    print(f"  • Kích thước: {comm_size} nodes ({comm_size/G_undirected.number_of_nodes()*100:.2f}%)")
    print(f"  • Loại nodes: {dict(node_types_in_comm)}")
    print(f"  • Top 3 nodes:")
    for node_id, degree in top_nodes_in_comm[:3]:
        node_title = nodes_df[nodes_df['id'] == node_id]['title'].values[0] if node_id in nodes_df['id'].values else node_id
        node_type = nodes_df[nodes_df['id'] == node_id]['type'].values[0] if node_id in nodes_df['id'].values else 'unknown'
        print(f"      - {node_title} [{node_type}] (Degree: {degree})")

# Thống kê cộng đồng
print(f"\n[THỐNG KÊ CỘNG ĐỒNG]")
print(f"Kích thước cộng đồng:")
sizes = [stat['size'] for stat in community_stats]
print(f"  • Lớn nhất: {max(sizes)}")
print(f"  • Nhỏ nhất: {min(sizes)}")
print(f"  • Trung bình: {np.mean(sizes):.1f}")
print(f"  • Trung vị: {np.median(sizes):.1f}")

## 5️⃣ Phát Hiện Cộng Đồng (Community Detection)

In [ ]:
# Hình ảnh hóa xếp hạng
print("\n[Hình ảnh hóa XẾP HẠNG]")

fig, axes = plt.subplots(2, 2, figsize=(16, 12))

# PageRank
pagerank_df = pd.DataFrame(list(top_pagerank), columns=['Node', 'Score'])
pagerank_df['Label'] = pagerank_df['Node'].apply(lambda x: nodes_df[nodes_df['id'] == x]['title'].values[0][:20] if x in nodes_df['id'].values else x)
axes[0, 0].barh(range(len(pagerank_df)), pagerank_df['Score'], color='#FF6B6B')
axes[0, 0].set_yticks(range(len(pagerank_df)))
axes[0, 0].set_yticklabels(pagerank_df['Label'], fontsize=9)
axes[0, 0].set_xlabel('PageRank Score', fontsize=10)
axes[0, 0].set_title('Top 10 Nodes - PageRank', fontsize=12, fontweight='bold')
axes[0, 0].invert_yaxis()

# Degree Centrality
degree_df = pd.DataFrame(list(top_degree), columns=['Node', 'Score'])
degree_df['Label'] = degree_df['Node'].apply(lambda x: nodes_df[nodes_df['id'] == x]['title'].values[0][:20] if x in nodes_df['id'].values else x)
axes[0, 1].barh(range(len(degree_df)), degree_df['Score'], color='#4ECDC4')
axes[0, 1].set_yticks(range(len(degree_df)))
axes[0, 1].set_yticklabels(degree_df['Label'], fontsize=9)
axes[0, 1].set_xlabel('Degree Centrality Score', fontsize=10)
axes[0, 1].set_title('Top 10 Nodes - Degree Centrality', fontsize=12, fontweight='bold')
axes[0, 1].invert_yaxis()

# Betweenness Centrality
betweenness_df = pd.DataFrame(list(top_betweenness), columns=['Node', 'Score'])
betweenness_df['Label'] = betweenness_df['Node'].apply(lambda x: nodes_df[nodes_df['id'] == x]['title'].values[0][:20] if x in nodes_df['id'].values else x)
axes[1, 0].barh(range(len(betweenness_df)), betweenness_df['Score'], color='#95E1D3')
axes[1, 0].set_yticks(range(len(betweenness_df)))
axes[1, 0].set_yticklabels(betweenness_df['Label'], fontsize=9)
axes[1, 0].set_xlabel('Betweenness Centrality Score', fontsize=10)
axes[1, 0].set_title('Top 10 Nodes - Betweenness Centrality', fontsize=12, fontweight='bold')
axes[1, 0].invert_yaxis()

# Closeness Centrality
closeness_df = pd.DataFrame(list(top_closeness), columns=['Node', 'Score'])
closeness_df['Label'] = closeness_df['Node'].apply(lambda x: nodes_df[nodes_df['id'] == x]['title'].values[0][:20] if x in nodes_df['id'].values else x)
axes[1, 1].barh(range(len(closeness_df)), closeness_df['Score'], color='#F38181')
axes[1, 1].set_yticks(range(len(closeness_df)))
axes[1, 1].set_yticklabels(closeness_df['Label'], fontsize=9)
axes[1, 1].set_xlabel('Closeness Centrality Score', fontsize=10)
axes[1, 1].set_title('Top 10 Nodes - Closeness Centrality', fontsize=12, fontweight='bold')
axes[1, 1].invert_yaxis()

plt.tight_layout()
plt.savefig('graph_out/ranking_algorithms.png', dpi=300, bbox_inches='tight')
print("✓ Lưu: graph_out/ranking_algorithms.png")
plt.show()

In [1]:
print("\n" + "=" * 80)
print("THUẬT TOÁN XẾP HẠNG ĐỒ THỊ (GRAPH RANKING)")
print("=" * 80)

# PageRank
print("\n[1. PageRank Ranking]")
pagerank_scores = nx.pagerank(G)
top_pagerank = sorted(pagerank_scores.items(), key=lambda x: x[1], reverse=True)[:10]

print("Top 10 nodes theo PageRank:")
for rank, (node_id, score) in enumerate(top_pagerank, 1):
    node_title = nodes_df[nodes_df['id'] == node_id]['title'].values[0] if node_id in nodes_df['id'].values else node_id
    node_type = nodes_df[nodes_df['id'] == node_id]['type'].values[0] if node_id in nodes_df['id'].values else 'unknown'
    print(f"  {rank:2d}. {node_title:<40} [{node_type:10s}] Score: {score:.6f}")

# Degree Centrality
print("\n[2. Degree Centrality Ranking]")
degree_centrality = nx.degree_centrality(G)
top_degree = sorted(degree_centrality.items(), key=lambda x: x[1], reverse=True)[:10]

print("Top 10 nodes theo Degree Centrality:")
for rank, (node_id, score) in enumerate(top_degree, 1):
    node_title = nodes_df[nodes_df['id'] == node_id]['title'].values[0] if node_id in nodes_df['id'].values else node_id
    node_type = nodes_df[nodes_df['id'] == node_id]['type'].values[0] if node_id in nodes_df['id'].values else 'unknown'
    degree = G.degree(node_id)
    print(f"  {rank:2d}. {node_title:<40} [{node_type:10s}] Score: {score:.6f} (Degree: {degree})")

# Betweenness Centrality
print("\n[3. Betweenness Centrality Ranking]")
print("  → Tính toán (có thể tốn thời gian)...")
betweenness_centrality = nx.betweenness_centrality(G)
top_betweenness = sorted(betweenness_centrality.items(), key=lambda x: x[1], reverse=True)[:10]

print("Top 10 nodes theo Betweenness Centrality:")
for rank, (node_id, score) in enumerate(top_betweenness, 1):
    node_title = nodes_df[nodes_df['id'] == node_id]['title'].values[0] if node_id in nodes_df['id'].values else node_id
    node_type = nodes_df[nodes_df['id'] == node_id]['type'].values[0] if node_id in nodes_df['id'].values else 'unknown'
    print(f"  {rank:2d}. {node_title:<40} [{node_type:10s}] Score: {score:.6f}")

# Closeness Centrality (cho largest component)
print("\n[4. Closeness Centrality Ranking]")
print("  → Tính toán cho thành phần liên thông lớn nhất...")
closeness_centrality = nx.closeness_centrality(largest_cc)
top_closeness = sorted(closeness_centrality.items(), key=lambda x: x[1], reverse=True)[:10]

print("Top 10 nodes theo Closeness Centrality:")
for rank, (node_id, score) in enumerate(top_closeness, 1):
    node_title = nodes_df[nodes_df['id'] == node_id]['title'].values[0] if node_id in nodes_df['id'].values else node_id
    node_type = nodes_df[nodes_df['id'] == node_id]['type'].values[0] if node_id in nodes_df['id'].values else 'unknown'
    print(f"  {rank:2d}. {node_title:<40} [{node_type:10s}] Score: {score:.6f}")


THUẬT TOÁN XẾP HẠNG ĐỒ THỊ (GRAPH RANKING)

[1. PageRank Ranking]


NameError: name 'nx' is not defined

## 4️⃣ Thuật Toán Xếp Hạng Đồ Thị (PageRank, Degree Centrality, Betweenness)

In [5]:
print("=" * 80)
print("CHỨNG MINH KHÁI NIỆM THẾ GIỚI NHỎ (SMALL WORLD)")
print("=" * 80)

print("\n[Phân tích kết nối cộng đồng]")

# Tìm connected components
if nx.is_connected(G_undirected):
    print("✓ Đồ thị là liên thông (connected)")
    largest_cc = G_undirected
    num_components = 1
else:
    print("! Đồ thị không liên thông (disconnected)")
    components = list(nx.connected_components(G_undirected))
    num_components = len(components)
    largest_cc = G_undirected.subgraph(max(components, key=len)).copy()
    print(f"  • Số thành phần liên thông: {num_components}")
    print(f"  • Thành phần lớn nhất: {len(largest_cc)} nodes ({len(largest_cc)/G_undirected.number_of_nodes()*100:.1f}%)")

print(f"\n[Tính toán khoảng cách ngắn nhất]")

# Tính average shortest path length cho largest connected component
if len(largest_cc) > 1:
    try:
        avg_path_length = nx.average_shortest_path_length(largest_cc)
        print(f"✓ Khoảng cách ngắn nhất trung bình: {avg_path_length:.4f}")
    except:
        avg_path_length = None
        print("! Không thể tính (có thể do đồ thị không liên thông)")
else:
    avg_path_length = 0
    print("Đồ thị quá nhỏ để tính")

# Tính diameter
try:
    diameter = nx.diameter(largest_cc)
    print(f"✓ Đường kính của mạng (diameter): {diameter}")
except:
    diameter = None
    print("! Không thể tính diameter")

# Tính clustering coefficient
print(f"\n[Hệ số gom cụm (Clustering Coefficient)]")

avg_clustering = nx.average_clustering(G_undirected)
print(f"✓ Hệ số gom cụm trung bình: {avg_clustering:.6f}")

# So sánh với random graph
print(f"\n[So sánh với Random Graph]")

n_nodes = G_undirected.number_of_nodes()
n_edges = G_undirected.number_of_edges()
random_prob = (2 * n_edges) / (n_nodes * (n_nodes - 1))

G_random = nx.erdos_renyi_graph(n_nodes, random_prob)

if nx.is_connected(G_random):
    random_avg_path = nx.average_shortest_path_length(G_random)
else:
    largest_component = max(nx.connected_components(G_random), key=len)
    G_random_cc = G_random.subgraph(largest_component).copy()
    random_avg_path = nx.average_shortest_path_length(G_random_cc)

random_avg_clustering = nx.average_clustering(G_random)

print(f"Alumni Network:")
print(f"  • Avg. Path Length: {avg_path_length:.4f}")
print(f"  • Avg. Clustering: {avg_clustering:.6f}")

print(f"\nRandom Graph (p={random_prob:.6f}):")
print(f"  • Avg. Path Length: {random_avg_path:.4f}")
print(f"  • Avg. Clustering: {random_avg_clustering:.6f}")

# Kiểm tra tính chất Small World
sigma = (avg_clustering / random_avg_clustering) / (avg_path_length / random_avg_path) if random_avg_path > 0 else 0
print(f"\n[Small World Coefficient σ = C/C_random ÷ L/L_random]")
print(f"✓ σ = {sigma:.4f}")

if sigma > 1:
    print("✓ Mạng này có tính chất THẾ GIỚI NHỎ (Small World)")
    print("  → Có gom cụm cao (high clustering)")
    print("  → Nhưng khoảng cách ngắn (short paths)")
else:
    print("! Mạng này KHÔNG có tính chất Small World")

CHỨNG MINH KHÁI NIỆM THẾ GIỚI NHỎ (SMALL WORLD)

[Phân tích kết nối cộng đồng]
! Đồ thị không liên thông (disconnected)
  • Số thành phần liên thông: 98
  • Thành phần lớn nhất: 2055 nodes (95.1%)

[Tính toán khoảng cách ngắn nhất]
  • Số thành phần liên thông: 98
  • Thành phần lớn nhất: 2055 nodes (95.1%)

[Tính toán khoảng cách ngắn nhất]
✓ Khoảng cách ngắn nhất trung bình: 3.3011
✓ Khoảng cách ngắn nhất trung bình: 3.3011
✓ Đường kính của mạng (diameter): 9

[Hệ số gom cụm (Clustering Coefficient)]
✓ Đường kính của mạng (diameter): 9

[Hệ số gom cụm (Clustering Coefficient)]
✓ Hệ số gom cụm trung bình: 0.456741

[So sánh với Random Graph]
✓ Hệ số gom cụm trung bình: 0.456741

[So sánh với Random Graph]
Alumni Network:
  • Avg. Path Length: 3.3011
  • Avg. Clustering: 0.456741

Random Graph (p=0.024309):
  • Avg. Path Length: 2.2477
  • Avg. Clustering: 0.024375

[Small World Coefficient σ = C/C_random ÷ L/L_random]
✓ σ = 12.7590
✓ Mạng này có tính chất THẾ GIỚI NHỎ (Small World)
  

## 3️⃣ Chứng Minh Khái Niệm Thế Giới Nhỏ (Small World)

In [4]:
# Xây dựng đồ thị NetworkX
print("\n[+] Xây dựng đồ thị...")

# Tạo graph từ edge list
G = nx.DiGraph()

# Thêm nodes
for _, row in nodes_df.iterrows():
    G.add_node(row['id'], title=row['title'], node_type=row['type'])

# Thêm edges
for _, row in edges_df.iterrows():
    G.add_edge(row['from'], row['to'], 
              relation_type=row['type'], 
              weight=row.get('weight', 1))

print(f"✓ Đồ thị xây dựng thành công")
print(f"  • Nodes: {G.number_of_nodes()}")
print(f"  • Edges: {G.number_of_edges()}")
print(f"  • Mật độ: {nx.density(G):.6f}")

# Chuyển sang undirected graph để phân tích cộng đồng
G_undirected = G.to_undirected()
print(f"\n✓ Undirected graph: {G_undirected.number_of_nodes()} nodes, {G_undirected.number_of_edges()} edges")


[+] Xây dựng đồ thị...
✓ Đồ thị xây dựng thành công
  • Nodes: 2162
  • Edges: 57975
  • Mật độ: 0.012409
✓ Đồ thị xây dựng thành công
  • Nodes: 2162
  • Edges: 57975
  • Mật độ: 0.012409

✓ Undirected graph: 2162 nodes, 56787 edges

✓ Undirected graph: 2162 nodes, 56787 edges


In [3]:
# Tải dữ liệu
print("[+] Đang tải dữ liệu...")

# Tải nodes
with open('graph_out/nodes_unified.csv', 'r', encoding='utf-8') as f:
    nodes_df = pd.read_csv(f)

# Tải edges
with open('graph_out/edges_unified.csv', 'r', encoding='utf-8') as f:
    edges_df = pd.read_csv(f)

print(f"✓ Đã tải {len(nodes_df)} nodes và {len(edges_df)} edges")

# Thống kê nodes theo loại
print("\n[NODES THEO LOẠI]")
node_types = nodes_df['type'].value_counts()
for node_type, count in node_types.items():
    print(f"  • {node_type}: {count}")

# Thống kê edges theo loại
print("\n[EDGES THEO LOẠI]")
edge_types = edges_df['type'].value_counts()
for edge_type, count in edge_types.items():
    print(f"  • {edge_type}: {count}")

[+] Đang tải dữ liệu...
✓ Đã tải 2162 nodes và 66910 edges

[NODES THEO LOẠI]
  • person: 1229
  • university: 842
  • country: 67
  • career: 24

[EDGES THEO LOẠI]
  • same_birth_country: 39957
  • link_to: 15319
  • same_uni: 8707
  • alumni_of: 1629
  • same_career: 1298


## 2️⃣ Tải và Chuẩn Bị Dữ Liệu Mạng

In [2]:
import json
import pandas as pd
import numpy as np
import networkx as nx
import matplotlib.pyplot as plt
import seaborn as sns
from collections import Counter, defaultdict
import warnings
warnings.filterwarnings('ignore')

# Thiết lập font cho Vietnamese
plt.rcParams['font.sans-serif'] = ['DejaVu Sans']
plt.rcParams['axes.unicode_minus'] = False

print("✓ Thư viện được import thành công")

✓ Thư viện được import thành công


## 1️⃣ Import Thư Viện Cần Thiết

# 📊 Phân Tích Mạng Xã Hội Alumni - Social Network Analysis

## Mục tiêu
Phân tích mạng alumni bằng các thuật toán nâng cao:
1. ✓ Chứng minh khái niệm **thế giới nhỏ** (Small World)
2. ✓ Xếp hạng nodes bằng **PageRank** và các thuật toán đồ thị
3. ✓ Phát hiện **cộng đồng** (Community Detection)
4. ✓ Phân tích **độ trung tâm** (Centrality Measures)

## Dữ liệu
- **Nodes**: 2,162 (người, đại học, quốc gia, nghề)
- **Edges**: 66,910 (tất cả các mối quan hệ)
- **Loại mối quan hệ**: same_birth_country, link_to, same_uni, alumni_of, same_career

In [ ]:
print("\n" + "=" * 80)
print("TÓM TẮT & KẾT LUẬN - PHÂN TÍCH MẠNG XÃ HỘI ALUMNI")
print("=" * 80)

print("\n[1. KHÁI NIỆM THẾ GIỚI NHỎ (SMALL WORLD)]")
print(f"  ✓ Khoảng cách ngắn nhất trung bình: {avg_path_length:.4f} hops")
if diameter:
    print(f"  ✓ Đường kính mạng: {diameter}")
print(f"  ✓ Hệ số gom cụm: {avg_clustering:.6f}")
print(f"  ✓ Small World Coefficient (σ): {sigma:.4f}")

if sigma > 1:
    print(f"\n  ➜ KẾT LUẬN: Mạng Alumni có tính chất THẾ GIỚI NHỎ")
    print(f"     - Người dùng cách nhau rất gần (avg path ≈ {avg_path_length:.2f})")
    print(f"     - Nhưng có độ gom cụm cao (clustering ≈ {avg_clustering:.4f})")
    print(f"     - Điều này cho thấy sự tồn tại của các \"nhóm bạn\" chặt chẽ")
else:
    print(f"\n  ➜ KẾT LUẬN: Mạng có tính chất không hoàn toàn Small World")

print("\n[2. THUẬT TOÁN XẾP HẠNG (RANKING ALGORITHMS)]")
print(f"  ✓ PageRank: Xác định các node có ảnh hưởng cao")
print(f"  ✓ Degree Centrality: Xác định các hub trực tiếp kết nối")
print(f"  ✓ Betweenness Centrality: Xác định những người \"cầu nối\"")
print(f"  ✓ Closeness Centrality: Xác định những người \"trung tâm\"")
print(f"\n  ➜ Top node theo PageRank:")

if top_pagerank:
    top_node = top_pagerank[0]
    top_title = nodes_df[nodes_df['id'] == top_node[0]]['title'].values[0] if top_node[0] in nodes_df['id'].values else top_node[0]
    top_type = nodes_df[nodes_df['id'] == top_node[0]]['type'].values[0] if top_node[0] in nodes_df['id'].values else 'unknown'
    print(f"     - {top_title} [{top_type}] (Score: {top_node[1]:.6f})")

print("\n[3. PHÁT HIỆN CỘNG ĐỒNG (COMMUNITY DETECTION)]")
print(f"  ✓ Số cộng đồng phát hiện: {len(communities_louvain)}")
print(f"  ✓ Modularity Score: {modularity_louvain:.6f}")
print(f"  ✓ Kích thước cộng đồng từ {min(sizes)} đến {max(sizes)} nodes")

print(f"\n  ➜ Ý NGHĨA:")
print(f"     - Mạng Alumni tự tổ chức thành {len(communities_louvain)} cộng đồng")
print(f"     - Mỗi cộng đồng đại diện một \"nhánh\" của mạng với kết nối chặt chẽ")
if modularity_louvain > 0.3:
    print(f"     - Modularity cao ({modularity_louvain:.4f}) → Cộng đồng rõ ràng, tách biệt")
elif modularity_louvain > 0.1:
    print(f"     - Modularity trung bình ({modularity_louvain:.4f}) → Cộng đồng yếu, nhiều liên kết xuyên")
else:
    print(f"     - Modularity thấp ({modularity_louvain:.4f}) → Mạng liên thông tốt, ít tách biệt")

print("\n" + "=" * 80)
print("✓ PHÂN TÍCH HOÀN TẤT")
print("=" * 80)

## 6️⃣ Tóm Tắt & Kết Luận

In [ ]:
# Hình ảnh hóa cộng đồng
print("\n[Hình ảnh hóa CỘNG ĐỒNG]")

fig, axes = plt.subplots(1, 2, figsize=(16, 6))

# Biểu đồ kích thước cộng đồng
comm_sizes = [stat['size'] for stat in community_stats]
comm_labels = [f"Community {i+1}" for i in range(len(community_stats))]
colors_comm = plt.cm.Set3(np.linspace(0, 1, len(comm_sizes)))

axes[0].bar(range(len(comm_sizes)), comm_sizes, color=colors_comm)
axes[0].set_xticks(range(len(comm_sizes)))
axes[0].set_xticklabels(comm_labels, rotation=45, ha='right')
axes[0].set_ylabel('Số Lượng Nodes', fontsize=11)
axes[0].set_title('Kích Thước Các Cộng Đồng', fontsize=12, fontweight='bold')
axes[0].grid(axis='y', alpha=0.3)

# Phân bố loại nodes trong mỗi cộng đồng
node_type_distribution = {}
for node_type in nodes_df['type'].unique():
    node_type_distribution[node_type] = [stat['node_types'].get(node_type, 0) for stat in community_stats]

x = np.arange(len(comm_labels))
width = 0.2
colors_types = {'person': '#FF6B6B', 'university': '#4ECDC4', 'country': '#95E1D3', 'career': '#F38181'}

for i, (node_type, values) in enumerate(node_type_distribution.items()):
    offset = (i - len(node_type_distribution) / 2) * width
    axes[1].bar(x + offset, values, width, label=node_type, color=colors_types.get(node_type, '#999999'))

axes[1].set_xticks(x)
axes[1].set_xticklabels(comm_labels, rotation=45, ha='right')
axes[1].set_ylabel('Số Lượng Nodes', fontsize=11)
axes[1].set_title('Phân Bố Loại Nodes Theo Cộng Đồng', fontsize=12, fontweight='bold')
axes[1].legend(fontsize=10)
axes[1].grid(axis='y', alpha=0.3)

plt.tight_layout()
plt.savefig('graph_out/community_detection.png', dpi=300, bbox_inches='tight')
print("✓ Lưu: graph_out/community_detection.png")
plt.show()